### PR TITLE
Note shot simulation

### LINKED ISSUE #28: Note shot simulation
Given speed and angle of bot and shooter, calculate and map out the trajectory the note will take

### PR DESCRIPTION
**Pull request by icai0**
simulates note shooting and trajectory given shooter values, later add so that it only shoots when there is a note held

### CODE REVIEW

File: ControlScheme.json [hunk 1]
```diff
@@ -110,5 +110,12 @@
       "Button": 1
     },
     "Replace"
+  ],
```
Comment (lmaxwell24) [inline ControlScheme.json]: merge with main for actual controls

File: commands/shooter/shootsimnote.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+from commands2 import Command
+from subsystems.shootersubsystem import ShooterSubsystem
+
+# temp command for testing note sim, general score command will be made later
+
+
+class ShootSimNote(Command):
```
Comment (lmaxwell24) [inline commands/shooter/shootsimnote.py]: merge this with the feed note, triggering simulation should be the same as note feed

File: constants.py [hunk 3]
```diff
@@ -822,3 +822,15 @@
...
 """Revolutions"""
+
+kRobotToShooterTransform = Transform3d(
+    Pose3d(),
+    Pose3d(
+        -5.631 * kMetersPerInch, 0, 13.191 * kMetersPerInch, Rotation3d(0, 0, math.pi)
+    ),
+)
+
+kShooterWheelDiameter = 3.965 * kMetersPerInch
```
Comment (lmaxwell24) [inline constants.py]: whats this wheel diameter from? I though they were 4in

File: operatorinterface.py [hunk 4]
```diff
@@ -190,3 +190,5 @@ def getPOVBindingOfName(name: str) -> typing.Tuple[Joystick, int, int]:
         self.trap = getButtonBindingOfName("trap")
         self.hold = getButtonBindingOfName("hold")
         self.feed = getButtonBindingOfName("feed")
+
+        self.simshoot = getButtonBindingOfName("simshoot")
```
Comment (lmaxwell24) [inline operatorinterface.py]: see comments about merging with feed

File: subsystems/shootersubsystem.py [hunk 5]
```diff
@@ -1,18 +1,59 @@
...
+            self.zc = self.zi + self.vz * elapsedTime - 9.8 / 2 * elapsedTime**2
+
+
 class ShooterSubsystem(Subsystem):
-    def __init__(self) -> None:
+    def __init__(self, timer: Timer) -> None:
         Subsystem.__init__(self)
         self.setName(__class__.__name__)
 
+        self.timer = timer
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: why can't a new timer object be made for shooter subsystem?

File: subsystems/shootersubsystem.py [hunk 6]
```diff
@@ -76,9 +117,11 @@ def __init__(self) -> None:
         # speeds are in RPM, same as velocity control mode
 
     def setShooterAngle(self, angle: Rotation2d) -> None:
-        self.targetAngle = min(
-            max(constants.kShooterMinAngle, angle),
-            constants.kShooterMaxAngle,
+        self.targetAngle = Rotation2d(
+            min(
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: use clamp in convenientmath

File: subsystems/shootersubsystem.py [hunk 7]
```diff
@@ -150,18 +193,98 @@ def readyToShoot(self) -> bool:
...
 
-    def periodic(self) -> None:
-        # logging
-        SmartDashboard.putNumber(
-            constants.kShooterAngleKey, self.getShooterAngle().radians()
+    def addSimNote(self) -> None:
+        pose = SmartDashboard.getNumberArray(
+            constants.kRobotPoseArrayKeys.valueKey, [0, 0, 0]
+        )
+        robotPose = Pose2d(pose[0], pose[1], pose[2])
```
Comment (lmaxwell24) [inline subsystems/shootersubsystem.py]: nitpick: alt writing
Pose2d(*pose)

**Review by lmaxwell24 — CHANGES_REQUESTED** [review]

File: constants.py (hunk 3 above)
Comment (icai0) [reply constants.py]: cad

Comment (lmaxwell24) [reply constants.py]: image
:)

Comment (icai0) [reply constants.py]: ur gaslighting me

Comment (lmaxwell24) [reply constants.py]: the andymark wheels in cad have a bit of a ramp on the sides, idk how accurate it is but that number isn't from the center wheel radius
treat it as a 4in wheel cuz the ttb wheels are 4in

**Review by lmaxwell24 — APPROVED** [review]
very cool sim stuff

File: constants.py [hunk 8]
```diff
@@ -822,3 +822,15 @@
 
 kPullDownBandLimit = 0.1
 """Revolutions"""
+
+kRobotToShooterTransform = Transform3d(
+    Pose3d(),
```
Comment (LandonBayer) [inline constants.py:827]: why this just empty

File: constants.py [hunk 9]
```diff
@@ -822,3 +822,15 @@
...
+    Pose3d(
+        -5.631 * kMetersPerInch, 0, 13.191 * kMetersPerInch, Rotation3d(0, 0, math.pi)
+    ),
+)
+
+kShooterWheelDiameter = 4 * kMetersPerInch
+kSimNoteArrayKey = "SimNoteArray"
+kLatestNoteTrajectoryKey = "LatestNoteTrajectory"
+kNoteTrajectoryTimeInterval = 0.15
```
Comment (LandonBayer) [inline constants.py:836]: Is this number arbitrary?

File: subsystems/shootersubsystem.py [hunk 10]
```diff
@@ -150,18 +195,98 @@ def readyToShoot(self) -> bool:
...
+        pose = SmartDashboard.getNumberArray(
+            constants.kRobotPoseArrayKeys.valueKey, [0, 0, 0]
+        )
+        robotPose = Pose2d(*pose)
+        robotVelocities = SmartDashboard.getNumberArray(
+            constants.kDriveVelocityKeys, [0, 0, 0]
+        )
+        shooterPose = Pose3d(robotPose) + constants.kRobotToShooterTransform
+
+        noteSpeed = (
```
Comment (LandonBayer) [inline subsystems/shootersubsystem.py:208]: is the unit for this in/sec?

**Review by LiamMcWeeney58 — APPROVED** [review]
hi

**Review by LandonBayer — APPROVED** [review]

File: constants.py (hunk 8 above)
Comment (lmaxwell24) [reply constants.py:827]: transforms can be generated between two poses

File: constants.py (hunk 9 above)
Comment (lmaxwell24) [reply constants.py:836]: yeah its for how fine precision the note trajectory should be generated

File: subsystems/shootersubsystem.py (hunk 10 above)
Comment (lmaxwell24) [reply subsystems/shootersubsystem.py:208]: m/s

File: constants.py (hunk 8 above)
Comment (LandonBayer) [reply constants.py:827]: So the first pose is just 0,0,0?

File: constants.py (hunk 9 above)
Comment (LandonBayer) [reply constants.py:836]: so lower is better but possibly laggier?

File: constants.py (hunk 8 above)
Comment (lmaxwell24) [reply constants.py:827]: its the transform to get from 0,0,0 to the other pose, or robot origin to the shooter launch location

File: constants.py (hunk 9 above)
Comment (lmaxwell24) [reply constants.py:836]: yep

File: constants.py (hunk 8 above)
Comment (LandonBayer) [reply constants.py:827]: So in sim 0,0,0 is the robot starting position, and therefore the transform to the shooter can be set from it. Makes sense

File: subsystems/shootersubsystem.py (hunk 10 above)
Comment (LandonBayer) [reply subsystems/shootersubsystem.py:208]: Ah wheel diameter converts from inches to meters within, got it.

**Pull request merged**